HGCAL end of life DIGITISATION update

franzoni opened the pull request:
#### PR description:

We've documented in detail the content of this PR in yesterday's HGCAL DPG meeting https://indico.cern.ch/event/931566/ 

1 . Electronics specifications updated from HGCROCV2 to V3
2 . Leakage of amplitude from BX-1 updated and made gain specific
3 . fluence updated to latest maps from bril
4 . logic to detect optimal HGCROCv3 gain slightly modified to favor the MIP peak in the [8-15] ADC range

All these changes won't affect the HGCAL DIGI in startup. 
They all affect only DIGI the end-of-life configuration ( ``SLHCUpgradeSimulations/Configuration/aging.customise_aging_3000 ``).


#### PR validation:

We've run workflow 23434 w/ and w/o customisation for ageing at 3000/fg. No significant changes in the startup. CPU time per event decreases in the 3000/fb scenario by a few seconds, driven by the reduction in hit occupancy

pfs committed:
- HGCalRadiationMap
  . deprecated radiiVec to give more freedom for different parameterizations (needed for the latest fluka map)
- HGcalSciNoiseMap and HGCHEbackSignalScalerAnalyzer
  . propagation of the depcreation of radiiVec
- HGCalSiNoiseMap
  . adding method to get Si operation characteristics independently of the DetId
  . adding the split of the noise components

pfs committed:
Adding newer parameterizations from fluka, the previous are still kept for reference on purpose

pfs committed:
- HGCHEbackDigitizer.cc
   . propagating deprecation of radiiVec

pfs committed:
- HGCalSiNoiseMap
 . removing HGCROCv2 backward compatibility
 . introducing a cache to avoild re-evaluating radius, fluence and all the parameters for Si

pfs committed:
tested cache procedure (unfortunately wafer equivalence can't be applied)

pfs committed:
updating fluka map in cfi

pfs committed:
. HGCalSiNoiseMap
  - added a core struct with the minimal set of variables used in the digitizer
  - added method to use the cache based on +z DetIds
. HGCSiNoiseMapAnalyzer and HGCDigitizerBase
  - propagated the change in api introduced above

pfs committed:
fixing default...

pfs committed:
- HGCalSiNoiseMap

  . introducing gain-dependent pulse shapes according to tables from D. Thienpont
  . adding two additional bits to dose algo to disable gain-dependent pulse shape and usage of cache

- HGCFEElectronics

  . adding method to retrieve default ADC pulse shape configured via phyton
  . adding optional adcPulse argument to run shaper methods

- HGCDigitizerBase

  . propagation of the changes above

Commit:
custom to turn off caching; and some doc

Commit:
custom to turn off caching; and some doc

Commit:
add customs to toggle algo, and fix the one for caching

Commit:
add more customs to toggle algo, and fix the noise

Commit:
spaces

Commit:
fix the HGCal_setRealisticStartupNoise

Commit:
do cache the siop's

pfs committed:
Merged hgcal_eol_pulse_update_112X_bis from repository franzoni with cms-merge-topic

pfs committed:
- HGCalSiNoiseMap
  setGeometry method now takes care of filling the operation cache
  cache filled from the list of valid DetIds from geometry insted of dynamically from DetIds received.
  This avoids having to query if DetId is already existing in the cache.
  Given there is no need to have a fast query on the cache, switched from unordered_map to a standard map
  to reduce memory footprint

- HGCDigitizerBase
  propagation of the changes above

Commit:
Merge pull request #14 from PFCal-dev/hgcal_eol_pulse_update_112X_bis

Hgcal eol pulse update 112 x bis

pfs committed:
HGCalSiNoiseMap

- disable cached op by default

pfs committed:
- HGCalSiNoiseMap
   . update pad area for latest sensor specs

Commit:
no cache as default for siop parameters

pfs committed:
changing the logic that sets the gain

pfs committed:
Merge branch 'hgcal_eol_pulse_update_112X_bis' into hgcal_eol_pulse_update_112X_bis

Commit:
Merge pull request #15 from PFCal-dev/hgcal_eol_pulse_update_112X_bis

. update surface of pads
. modify logic of auto-gain

Commit:
bump up gain of pas around 800mm

Commit:
bump up gain of pas around 800mm

Commit:
code-checks

Commit:
code-format

franzoni commented:
please test

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30422/16491

- This PR adds an extra 216KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
A new Pull Request was created by  @franzoni (Giovanni Franzoni) for master.

It involves the following packages:

SimCalorimetry/HGCalSimAlgos
SimCalorimetry/HGCalSimProducers


@cmsbuild, @civanch, @kpedro88, @mdhildreth can you please review it and eventually sign? Thanks.
@vandreev11, @sethzenz, @makortel, @rovere, @lgray, @cseez, @apsallid, @pfs, @deguio, @hatakeyamak this is something you requested to watch as well.
@silviodonato, @dpiparo you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
+1
Tested at: 9014602977c0916ab23b12e9293550a85b1f2711
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-716ddc/7405/summary.html
CMSSW: CMSSW_11_2_X_2020-06-25-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-716ddc/7405/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2778915
 - DQMHistoTests: Total failures: 4
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2778861
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

pfs committed:
Merged hgcal_eol_pulse_update_112X_bis from repository franzoni with cms-merge-topic

rovere commented on SimCalorimetry/HGCalSimAlgos/interface/HGCalRadiationMap.h:
```diff
@@ -7,36 +7,38 @@
-  double getFluenceValue(const int, const int, const radiiVec &, bool logVal = false);
+  double computeRadius(const HGCScintillatorDetId &);
+
+  double getDoseValue(const int, const int, const double &, bool logVal = false);
```
What is the rationale of passing a `const double &`?
The `&` was there before since the passed type was an `std::array`.

franzoni replied:
ok

rovere commented on SimCalorimetry/HGCalSimAlgos/interface/HGCalRadiationMap.h:
```diff
@@ -7,36 +7,38 @@
+  double computeRadius(const HGCScintillatorDetId &);
+
+  double getDoseValue(const int, const int, const double &, bool logVal = false);
+  double getFluenceValue(const int, const int, const double &, bool logVal = false);
```
Same comment applies here, too.

rovere commented on SimCalorimetry/HGCalSimAlgos/interface/HGCalSciNoiseMap.h:
```diff
@@ -18,18 +18,18 @@ class HGCalSciNoiseMap : public HGCalRadiationMap {
-  double scaleByTileArea(const HGCScintillatorDetId&, const radiiVec&);
-  double scaleBySipmArea(const HGCScintillatorDetId&, const double&);
-  std::pair<double, double> scaleByDose(const HGCScintillatorDetId&, const radiiVec&);
+  double scaleByTileArea(const HGCScintillatorDetId &, const double &);
```
Same here and in the following declarations

franzoni replied:
ok

pfs committed:
Addressing some comments

https://github.com/cms-sw/cmssw/pull/30422#discussion_r4456831141 changed comment in the code, the function specializes, does not override in practice

https://github.com/cms-sw/cmssw/pull/30422#discussion_r445684107 changed to const

https://github.com/cms-sw/cmssw/pull/30422#discussion_r445684907 done

https://github.com/cms-sw/cmssw/pull/30422#discussion_r445685230 the unordered_map adds some non-negligible extra mem to the process. We checked that the std::map will still perform faster per event than re-computing every single event as we don't need to query anywhere the map (find, count ,etc.) , just need to access it to retrieve the required values

https://github.com/cms-sw/cmssw/pull/30422#discussion_r445685639 defined a new type hgc_digi::FEPulseShape in HGCFEElectronics.h

https://github.com/cms-sw/cmssw/pull/30422#discussion_r445701175 +
https://github.com/cms-sw/cmssw/pull/30422#discussion_r445701446 +
https://github.com/cms-sw/cmssw/pull/30422#discussion_r445701776 +
https://github.com/cms-sw/cmssw/pull/30422#discussion_r445702370 changed accordingly

rovere commented on SimCalorimetry/HGCalSimProducers/python/hgcalDigitizer_cfi.py:
```diff
@@ -233,17 +283,15 @@ def HGCal_setRealisticNoiseSi(process,byDose=True,byDoseAlgo=0):
     return process
 
-def HGCal_setRealisticNoiseSci(process,byDose=True,byDoseAlgo=0):
+def HGCal_setRealisticNoiseSci(process,byDose=True,byDoseAlgo=0,byDoseMap=cms.string("SimCalorimetry/HGCalSimProducers/data/doseParams_3000fb_fluka-3.7.20.txt")):
```
I'm not sure spreading the `byDoseMap` as default parameter in so many different places is desirable.
Wouldn't it better to define it once and reference it everywhere else?

franzoni replied:
ok

franzoni commented on SimCalorimetry/HGCalSimAlgos/interface/HGCalSiNoiseMap.h:
```diff
@@ -62,36 +91,52 @@ class HGCalSiNoiseMap : public HGCalRadiationMap {
   std::vector<std::vector<double> > &getENCsParam() { return encsParam_; }
   std::vector<double> &getLSBPerGain() { return lsbPerGain_; }
+  void setDefaultADCPulseShape(const std::array<float, 6> &adcPulse) { defaultADCPulse_ = adcPulse; }
+  const std::array<float, 6> &getADCPulseForGain(GainRange_t gain) {
```
ok, we'll do both

franzoni commented on SimCalorimetry/HGCalSimAlgos/interface/HGCalSiNoiseMap.h:
```diff
@@ -62,36 +91,52 @@ class HGCalSiNoiseMap : public HGCalRadiationMap {
+  int defaultAimMIPtoADC_;
+
+  //cache of SiCellOpCharacteristics
+  std::map<uint32_t, SiCellOpCharacteristicsCore> siopCache_;
```
We're not using the siopCache_ at the moment because of the memory it would add (see our DPG talk): 50+ MB per core. If we were to use the caching, we'd want to keep its memory to the minimum  (hence ``std::map``) , while the computation which will be saved will vastly exceed the lower search performance with respect to ``std::unordered_map``

We keep the caching code available because in the next version of the geometry we'll be able to fold the detector and slash the memory by a factor 3 or even 6

franzoni commented on SimCalorimetry/HGCalSimAlgos/interface/HGCalSiNoiseMap.h:
```diff
@@ -62,36 +91,52 @@ class HGCalSiNoiseMap : public HGCalRadiationMap {
-  //electronics noise (series+parallel) polynomial coeffs;
+  //electronics noise (series+parallel) polynomial coeffs and ADC pulses;
   std::vector<std::vector<double> > encsParam_;
+  std::array<float, 6> defaultADCPulse_;
```
ok.

franzoni commented on SimCalorimetry/HGCalSimAlgos/src/HGCalSiNoiseMap.cc:
```diff
@@ -50,91 +62,201 @@ void HGCalSiNoiseMap::setDoseMap(const std::string &fullpath, const unsigned int
 }
 
+//
+double HGCalSiNoiseMap::getENCpad(const double &ileak) {
```
ok

franzoni commented on SimCalorimetry/HGCalSimAlgos/src/HGCalSiNoiseMap.cc:
```diff
@@ -50,91 +62,201 @@ void HGCalSiNoiseMap::setDoseMap(const std::string &fullpath, const unsigned int
+    return;
+
+  const std::vector<DetId> &validDetIds = geom()->getValidDetIds();
+  for (auto &did : validDetIds) {
```
ok

kpedro88 replied:
not done?

franzoni replied:
ok

franzoni commented on SimCalorimetry/HGCalSimAlgos/src/HGCalSiNoiseMap.cc:
```diff
@@ -50,91 +62,201 @@ void HGCalSiNoiseMap::setDoseMap(const std::string &fullpath, const unsigned int
+    //compute and store in cache
+    SiCellOpCharacteristicsCore siop = getSiCellOpCharacteristicsCore(hgcDetId);
+    std::pair<uint32_t, SiCellOpCharacteristicsCore> toAdd(rawId, siop);
+    siopCache_.insert(toAdd);
```
ok

franzoni commented on SimCalorimetry/HGCalSimAlgos/src/HGCalSiNoiseMap.cc:
```diff
@@ -50,91 +62,201 @@ void HGCalSiNoiseMap::setDoseMap(const std::string &fullpath, const unsigned int
+  //location of the cell
+  int subdet(cellId.subdet());
+  std::vector<double> &cceParam = cceParam_[cellThick];
+  auto xy(
```
ok

franzoni commented on SimCalorimetry/HGCalSimAlgos/src/HGCalSiNoiseMap.cc:
```diff
@@ -50,91 +62,201 @@ void HGCalSiNoiseMap::setDoseMap(const std::string &fullpath, const unsigned int
+      gain = igain;
+    }
+
+    //previous algo (kept commented for the moment)
```
ok

Commit:
Merge pull request #16 from PFCal-dev/hgcal_eol_pulse_update_112X_bis

Addressing some comments by Kevin

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30422/16513

- This PR adds an extra 28KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30422/16513/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30422/16513/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

franzoni commented on SimCalorimetry/HGCalSimProducers/data/doseParams_3000fb_fluka-3.7.20.txt:
```diff
@@ -0,0 +1,64 @@
+8  1  0         0           0            0            0            0 15.616705     -0.0127294      -0.000138386    2.35373e-06 -1.05516e-08 40
```
these are 9kB.
In "good citizen" spirit, we'll remove the old version of the same content, which no longer has reason to be in release.

Commit:
removing unnecessary old doseParams_3000fb_fluka-3.5.15.9.txt

Commit:
rovere - What is the rationale of passing a const double &

cmsbuild commented:
The code-checks are being triggered in jenkins.

Commit:
code-checks code-format

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30422/16515

- This PR adds an extra 140KB to repository

cmsbuild commented:
Pull request #30422 was updated. @cmsbuild, @civanch, @kpedro88, @mdhildreth can you please check and sign again.

franzoni commented on SimCalorimetry/HGCalSimAlgos/interface/HGCalSiNoiseMap.h:
```diff
@@ -47,13 +54,35 @@ class HGCalSiNoiseMap : public HGCalRadiationMap {
+     cache the first time it is called - intrinsically related to the valid detIds in the geometry
+     the filling of the cache is ignored by configuration or if it has already been filled
+   */
+  void setGeometry(const CaloSubdetectorGeometry *, GainRange_t gain = GainRange_t::AUTO, int aimMIPtoADC = 10);
```
signatures are different. there's no ``override``-ing

Commit:
remove unneeded custom functions

cmsbuild commented:
The code-checks are being triggered in jenkins.

Commit:
code-format and code-checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

franzoni commented:
Thanks @kpedro88 and @rovere  for your feedback.
We've addressed all of them.

franzoni commented:
please test

cmsbuild commented:
-code-checks

ERROR: Build errors found during clang-tidy run.
```
gmake: *** [config/SCRAM/GMake/Makefile.coderules:128: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

kpedro88 commented on SimCalorimetry/HGCalSimAlgos/src/#HGCalSiNoiseMap.cc#:
```diff
@@ -0,0 +1,244 @@
+#include "SimCalorimetry/HGCalSimAlgos/interface/HGCalSiNoiseMap.h"
+
+//
```
please remove this file "#HGCalSiNoiseMap.cc#" from the PR entirely (including from the history)

franzoni replied:
Hello @kpedro88 , I seem not to be able to remove this file from the history.

I've tried this approach https://blog.liplex.de/remove-file-from-git-history/ 
(in a few variants; the process would take a few days to complete) as well as 
manually  rolling back the history+redoing the commit which accidentally added the file + cherry-pick all following commits 
(leading to "Updates were rejected because the tip of your current branch is behind" when pushing to github)

I am out of ideas. Do you have tips ?
Simply removing the file it's trivial and can be done right away
Thanks!

Commit:
rovere Same here and in the following declarations

Commit:
kpedro88 const auto&

cmsbuild commented:
The code-checks are being triggered in jenkins.

franzoni commented:
removed unnecessary file from the history ( thanks @kpedro88 for tips) and done the "not done?"

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30422/16538

- This PR adds an extra 40KB to repository

franzoni commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #30422 was updated. @cmsbuild, @civanch, @kpedro88, @mdhildreth can you please check and sign again.

cmsbuild commented:
+1
Tested at: 1f3b6f990306814ed2c3d7bf230abe4ea836cc24
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-716ddc/7446/summary.html
CMSSW: CMSSW_11_2_X_2020-06-26-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-716ddc/7446/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2778915
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2778864
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

civanch commented:
+1

@franzoni , I agree with Kevin, that this type of data should not be inside sub-directory but is an external data (but definition - it come our of external source and has non-negligible size).

franzoni commented:
hello @kpedro88 @civanch , what is the policy for files to go to external ? If appears a little bit of an overkill to us, it's a few kB. 

However, if the policy is this file ought to go to external, we will do that. In which case, 
could you please remind us of how to do it, perhaps there are instructions ?

It would be very helpful for us if this PR could make pre2.
Can we work on moving to the external after this PR ? It would be great if this PR could arrived signed at tomorrow's ORP...

rovere commented:
9kB is definitely not a huge penalty in terms of space. Space that, in any case, we have to pay through the externals anyway. So, I'm really questioning what the added value is to have it in a separate repository?
Given the fact that this PR is **on the critical path** to perform detailed back-end related studies, `HGCAL DPG` kindly asks to push it through as is in release ASAP. Any move to external repository or, even better, to real conditions, will be kept on our todo list and not forgotten.

kpedro88 commented:
@franzoni keeping data files in cms-data repositories makes it easier to clone the cmssw repository for development (otherwise the repository size can grow out of control)
but in this case it's not crucial

kpedro88 commented:
+upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
